### ISSUE TITLE
Avoid inducing latency in apply path due to schema version requesting

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#2109]
After schema changes on coordinator node, schema change notification processing on replicas races with mutations using the new schema. Notification will typically loose, because schema merging on replica node takes quite some time and new version is published only after that's done. Until that happens, mutations using the new schema will block on replica side on requesting schema version from the coordinator.

This blocking could be avoided if we ensure that definitions of new schema versions are sent to replicas before any mutation using them, leveraging the fact that the whole system is just a bunch of queues and queues don't reorder. 

TCP guarantees no reordering within a session. It would be possible (if it's not the case already) to ensure that RPC doesn't reorder calls within a session. For reconnects, we could allow registering a callback in RPC layer which would be executed when new session is established. So on reconnect, Scylla would clear the "already sent" version map, because we don't know if the versions we sent have reached the replica or not, and would resent them next time mutation is about to be sent.

**Comment by tgrabiec** [comment]
That would actually help only with reads, because writes would still have to wait for the schema merge to establish the latest version.

**Comment by tgrabiec** [comment]
The real fix would be to ensure we don't ACK the schema change (and thus start generating mutations using the new schema) until the change is merged on all live nodes (with a timeout), requiring a quorum for the operation to succeed.

**Comment by tgrabiec** [comment]
And that's what #1207 would achieve.

**Comment by gleb-cloudius** [comment]
On Thu, Feb 23, 2017 at 02:46:51AM -0800, Tomasz Grabiec wrote:
> After schema changes on coordinator node, schema change notification processing on replicas races with mutations using the new schema. Notification will typically loose, because schema merging on replica node takes quite some time and new version is published only after that's done. Until that happens, mutations using the new schema will block on replica side on requesting schema version from the coordinator.
> 
> This blocking could be avoided if we ensure that definitions of new schema versions are sent to replicas before any mutation using them, leveraging the fact that the whole system is just a bunch of queues and queues don't reorder. 
> 
How will this help given that "schema merging on replica
node takes quit some time"?

> TCP guarantees no reordering within a session. It would be possible (if it's not the case already) to ensure that RPC doesn't reorder calls within a session. For reconnects, we could allow registering a callback in RPC layer which would be executed when new session is established. So on reconnect, Scylla would clear the "already sent" version map, because we don't know if the versions we sent have reached the replica or not, and would resent them next time mutation is about to be sent.
> 
RPC does not reorder, but it executes calls asynchronously, so mutation
queued after "new schema" message will not wait for it to be processed
and may be reordered by the reactor.

--
			Gleb.

**Comment by tgrabiec** [comment]
2017-02-23 12:03 GMT+01:00 Gleb Natapov <notifications@github.com>:

> On Thu, Feb 23, 2017 at 02:46:51AM -0800, Tomasz Grabiec wrote:
> > After schema changes on coordinator node, schema change notification
> processing on replicas races with mutations using the new schema.
> Notification will typically loose, because schema merging on replica node
> takes quite some time and new version is published only after that's done.
> Until that happens, mutations using the new schema will block on replica
> side on requesting schema version from the coordinator.
> >
> > This blocking could be avoided if we ensure that definitions of new
> schema versions are sent to replicas before any mutation using them,
> leveraging the fact that the whole system is just a bunch of queues and
> queues don't reorder.
> >
> How will this help given that "schema merging on replica
> node takes quit some time"?
>
> > TCP guarantees no reordering within a session. It would be possible (if
> it's not the case already) to ensure that RPC doesn't reorder calls within
> a session. For reconnects, we could allow registering a callback in RPC
> layer which would be executed when new session is established. So on
> reconnect, Scylla would clear the "already sent" version map, because we
> don't know if the versions we sent have reached the replica or not, and
> would resent them next time mutation is about to be sent.
> >
> RPC does not reorder, but it executes calls asynchronously, so mutation
> queued after "new schema" message will not wait for it to be processed
> and may be reordered by the reactor.
>

new schema message could be processed without deferring.

**Comment by gleb-cloudius** [comment]
On Thu, Feb 23, 2017 at 03:12:17AM -0800, Tomasz Grabiec wrote:
> 2017-02-23 12:03 GMT+01:00 Gleb Natapov <notifications@github.com>:
> 
> > On Thu, Feb 23, 2017 at 02:46:51AM -0800, Tomasz Grabiec wrote:
> > > After schema changes on coordinator node, schema change notification
> > processing on replicas races with mutations using the new schema.
> > Notification will typically loose, because schema merging on replica node
> > takes quite some time and new version is published only after that's done.
> > Until that happens, mutations using the new schema will block on replica
> > side on requesting schema version from the coordinator.
> > >
> > > This blocking could be avoided if we ensure that definitions of new
> > schema versions are sent to replicas before any mutation using them,
> > leveraging the fact that the whole system is just a bunch of queues and
> > queues don't reorder.
> > >
> > How will this help given that "schema merging on replica
> > node takes quit some time"?
> >
> > > TCP guarantees no reordering within a session. It would be possible (if
> > it's not the case already) to ensure that RPC doesn't reorder calls within
> > a session. For reconnects, we could allow registering a callback in RPC
> > layer which would be executed when new session is established. So on
> > reconnect, Scylla would clear the "already sent" version map, because we
> > don't know if the versions we sent have reached the replica or not, and
> > would resent them next time mutation is about to be sent.
> > >
> > RPC does not reorder, but it executes calls asynchronously, so mutation
> > queued after "new schema" message will not wait for it to be processed
> > and may be reordered by the reactor.
> >
> 
> new schema message could be processed without deferring.
> 
Does it involve any cross shard operations or potential IO?

BTW what about other shard getting a mutation and sending it to a shard
that not yet has an up-to-date schema?

--
			Gleb.

**Comment by tgrabiec** [comment]
2017-02-23 12:18 GMT+01:00 Gleb Natapov <notifications@github.com>:

> On Thu, Feb 23, 2017 at 03:12:17AM -0800, Tomasz Grabiec wrote:
> > 2017-02-23 12:03 GMT+01:00 Gleb Natapov <notifications@github.com>:
> >
> > > On Thu, Feb 23, 2017 at 02:46:51AM -0800, Tomasz Grabiec wrote:
> > > > After schema changes on coordinator node, schema change notification
> > > processing on replicas races with mutations using the new schema.
> > > Notification will typically loose, because schema merging on replica
> node
> > > takes quite some time and new version is published only after that's
> done.
> > > Until that happens, mutations using the new schema will block on
> replica
> > > side on requesting schema version from the coordinator.
> > > >
> > > > This blocking could be avoided if we ensure that definitions of new
> > > schema versions are sent to replicas before any mutation using them,
> > > leveraging the fact that the whole system is just a bunch of queues and
> > > queues don't reorder.
> > > >
> > > How will this help given that "schema merging on replica
> > > node takes quit some time"?
> > >
> > > > TCP guarantees no reordering within a session. It would be possible
> (if
> > > it's not the case already) to ensure that RPC doesn't reorder calls
> within
> > > a session. For reconnects, we could allow registering a callback in RPC
> > > layer which would be executed when new session is established. So on
> > > reconnect, Scylla would clear the "already sent" version map, because
> we
> > > don't know if the versions we sent have reached the replica or not, and
> > > would resent them next time mutation is about to be sent.
> > > >
> > > RPC does not reorder, but it executes calls asynchronously, so mutation
> > > queued after "new schema" message will not wait for it to be processed
> > > and may be reordered by the reactor.
> > >
> >
> > new schema message could be processed without deferring.
> >
> Does it involve any cross shard operations or potential IO?
>

Processing of schema version entries doesn't require those, it only
installs the schema object in the per-shard schema_registry. This object
can then be picked up by later service::get_schema_definition() calls, so
reads for instance can proceed without extra blocking.

Writes are more complicated because they need to be applied locally only to
tables which have at least as recent schema as that of the incoming
mutation to avoid data loss. If their schema version obtained from
schema_registry is not marked as "synced", and initially it will not, it
means we don't yet know which version is latest, so they will additionally
block on the local schema merge to establish the latest schema among the
current schema of the node and the unsynced schema. This blocking could be
also avoided, by ensuring that schema change is ACKed and new schema is
used only after all live nodes synced with it, and that's #1207.



>
> BTW what about other shard getting a mutation and sending it to a shard
> that not yet has an up-to-date schema?


For propagating schema_ptr's across shards we use global_schema_ptr, we
already do that.

**Comment by tgrabiec** [comment]
Reopening, since #1207 doesn't ensure that all nodes have the same schema.

**Comment by tgrabiec** [comment]
Another thing we could do is to make schema propagation two-phase. Each table will now have two schema versions: replica-side schema and coordinator-side schema. The coordinator-side schema lags behind the replica-side schema.

Schema tables reflect the replica-side schema. On boot, coordinator-side schema is set to replica-side schema.

coordinator-side schema is used by the CQL layer. Replica-side schema is used by storage (`database` and below). 

Schema changes will be 2-step:
  1) send schema change mutations - this changes the replica-side schema.
  2) when all nodes ACK, tell them to switch their coordinator-side schema to the replica-side schema
  3) ACK the statement

This ensures that before coordinators start using the new schema, all replicas already know it and thus will not be hurt by schema merge latency.

Currently, the driver waits for schema agreement before ACKing the CQL schema change statement. If we employ this algorithm, this waiting will be now also inside Scylla. We could retire driver-side waiting later.

To make sure progress in case the schema change coordinator fails, nodes will switch their coordinator-side schema to the replica-side schema after a timeout.

**Comment by avikivity** [comment]
To avoid on-disk changes, perhaps something like this:

1. DDL coordinator: fellow coordinators, keep using schema version v1 for the next five seconds, or until I tell you
2. other coordinators: take a 5-second lock around source of schema used for coordinator
3. other coordinators: okay
4. DDL coordinator: perform schema change
5. DDL coordinator: wait for propagation
6. DDL coordinator: fellow coordinators, you may now use the new schema
7. other coordinators: release lock
8. other coordinators: okay

We'll need table::coordinator_schema() to provide the maybe-locked schema.

**Comment by tgrabiec** [comment]
> To avoid on-disk changes, perhaps something like this:
> 
> 1. DDL coordinator: fellow coordinators, keep using schema version v1 for the next five seconds, or until I tell you
> 2. other coordinators: take a 5-second lock around source of schema used for coordinator
> 3. other coordinators: okay
> 4. DDL coordinator: perform schema change
> 5. DDL coordinator: wait for propagation
> 6. DDL coordinator: fellow coordinators, you may now use the new schema
> 7. other coordinators: release lock
> 8. other coordinators: okay
> 
> We'll need table::coordinator_schema() to provide the maybe-locked schema.

This is very similar to this proposal: https://github.com/scylladb/scylla/issues/2109#issuecomment-840041635

**Comment by avikivity** [comment]
Yes, my goal was to propose how to do it without having two schema versions on disk.

**Comment by tgrabiec** [comment]
On Thu, May 13, 2021 at 11:57 AM Avi Kivity ***@***.***>
wrote:

> Yes, my goal was to propose how to do it without having two schema
> versions on disk.
>
I did not intend to have two schema versions on disk. Only replica-side
schema would be on disk.

>

**Comment by tgrabiec** [comment]
After https://github.com/scylladb/scylladb/commit/c16e4a04239798c41fa140c14eabb250b5422983 this problem will also impact reads.

**Comment by mykaul** [comment]
@avikivity , @tgrabiec - I'm setting this right now for 5.2, so we can agree on priority and design for a solution if one needed for this. We also need an assignee for this, if it's agreed to pursue.
If not, I suggest closing it and not leaving it rot in the backlog - if we'll hit this again, we can always re-open.

**Comment by tgrabiec** [comment]
What's the problem with keeping it open? If we close this, how do we distinguish this unimplemented idea from other issues, which were resolved?

**Comment by avikivity** [comment]
And we're hitting it every time we change schema; it's just usually not too terrible.

**Comment by mykaul** [comment]
> What's the problem with keeping it open? If we close this, how do we distinguish this unimplemented idea from other issues, which were resolved?

Closing means we've reached a decision on it, keeping it open with some futuristic version - less so.
I don't have a strong opinion about it, but I think it's better to keep a reasonable, realistic backlog.
This one is a good example - open for 5 years, with no clear action item. Again - we can keep it open if we feel there's value (for an on-going re-evaluation, link to newer issues as a 'meta-tracker', etc.)

As for distinguishing - there is more than one way of closing, so you can distinguish between 'completed' and 'not going to'.

**Comment by avikivity** [comment]
There's "Close as completed" and "Close as not planned" but neither fit the state of the issue.

**Comment by tgrabiec** [comment]
On Sun, Nov 20, 2022 at 5:10 PM Yaniv Kaul ***@***.***> wrote:

> What's the problem with keeping it open? If we close this, how do we
> distinguish this unimplemented idea from other issues, which were resolved?
>
> Closing means we've reached a decision on it, keeping it open with some
> futuristic version - less so.
> I don't have a strong opinion about it, but I think it's better to keep a
> reasonable, realistic backlog.
> This one is a good example - open for 5 years, with no clear action item.
> Again - we can keep it open if we feel there's value (for an on-going
> re-evaluation, link to newer issues as a 'meta-tracker', etc.)
>
There is an implementation plan in one of the comments (
https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635),
and we're hitting issues in production which this issue would help with
(linked). This issue is not a good example, in my opinion.

**Comment by mykaul** [comment]
> On Sun, Nov 20, 2022 at 5:10 PM Yaniv Kaul ***@***.***> wrote: What's the problem with keeping it open? If we close this, how do we distinguish this unimplemented idea from other issues, which were resolved? Closing means we've reached a decision on it, keeping it open with some futuristic version - less so. I don't have a strong opinion about it, but I think it's better to keep a reasonable, realistic backlog. This one is a good example - open for 5 years, with no clear action item. Again - we can keep it open if we feel there's value (for an on-going re-evaluation, link to newer issues as a 'meta-tracker', etc.)
> There is an implementation plan in one of the comments ( [#2109 (comment)](https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635)), and we're hitting issues in production which this issue would help with (linked). This issue is not a good example, in my opinion. Message ID: ***@***.***>
> […](#)

There is an implementation plan, which seems to be agreed on. Do we have an agreement on the priority (urgency) of implementing it?
I'm OK with not closing it, just unsure what purpose it'll serve. Shall I park it under 6.x? 5.x?

**Comment by gleb-cloudius** [comment]
IMO there shouldn't be any plan that does not involve raft here. So as far as I am concerned there is no plan.

**Comment by tgrabiec** [comment]
This affects only the "state machine" part of group0.

**Comment by gleb-cloudius** [comment]
> Why? Raft is orthogonal to this problem.

How is raft orthogonal to the problem? Quoting the issue: "leveraging the fact that the whole system is just a bunch of queues and queues don't reorder." Well raft will use completely different set of queues to begin with. Another issue is that with raft there is no longer ACKs from all replicas like it was in the old schema push code.

**Comment by tgrabiec** [comment]
@gleb-cloudius There are two ideas presented here. Remember that there are two kinds of schema pull, one which changes the node's local schema, and one which only pulls the definitions which live as a schema version in memory. The first idea with the queues does not plug into how schema changes are made, it only affects the request path so that it doesn't have to pull the schema version (in-memory one). Each request can use a different version of the schema. E.g. before sending a write request to some replica, we would send relevant schema version to that replica first, if it wasn't already. That doesn't involve merging with schema tables, so is relatively fast.

But before doing that, I think we should do the other proposal mentioned in this issue, which hides the latency of the heavy schema pull which merges schema definition with schema tables: https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635

This can also be implemented outside raft, and should work with both old and new mechanism.

**Comment by gleb-cloudius** [comment]
> @gleb-cloudius There are two ideas presented here. Remember that there are two kinds of schema pull, one which changes the node's local schema, and one which only pulls the definitions which live as a schema version in memory. The first idea with the queues does not plug into how schema changes are made, it only affects the request path so that it doesn't have to pull the schema version (in-memory one). Each request can use a different version of the schema. E.g. before sending a write request to some replica, we would send relevant schema version to that replica first, if it wasn't already.

How do you know it was not? It is sent over raft. You do not know who got it and who did not (without asking).

>  That doesn't involve merging with schema tables, so is relatively fast.

> 
> But before doing that, I think we should do the other proposal mentioned in this issue, which hides the latency of the heavy schema pull which merges schema definition with schema tables: [#2109 (comment)](https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635)
> 
> This can also be implemented outside raft, and should work with both old and new mechanism.

So you propose to do a global barrier (contact all nods) after the raft. Fine, but them it has the same issue as waiting in the drover. Unless you require all nodes to be alive during schema changed there will be races.

**Comment by tgrabiec** [comment]
On Tue, Nov 22, 2022 at 2:36 PM Gleb Natapov ***@***.***>
wrote:

> On Mon, Nov 21, 2022 at 1:17 PM Gleb Natapov *@*.***> wrote:
> @gleb-cloudius <https://github.com/gleb-cloudius>
> https://github.com/gleb-cloudius There are two ideas presented here.
> Remember that there are two kinds of schema pull, one which changes the
> node's local schema, and one which only pulls the definitions which live as
> a schema version in memory. The first idea with the queues does not plug
> into how schema changes are made, it only affects the request path so that
> it doesn't have to pull the schema version (in-memory one). Each request
> can use a different version of the schema. E.g. before sending a write
> request to some replica, we would send relevant schema version to that
> replica first, if it wasn't already. How do you know it was not? It is sent
> over raft. You do not know who got it and who did not (without asking).
>
> We can keep track of it in memory.
>
> Not without access to the raft internals which I am firmly against of. And
> even then the raft not always knows (but that involves leader change).
>
I think we don't need to touch raft internals, all can happen above it.
Firstly, the mapping doesn't have to be accurate, worst case the replica
will request the schema version like today. So we don't even need to update
the mapping when a schema change is made. The read/write coordinator will
push the schema version to target replicas with the first request which
uses the new version. That part doesn't involve raft.

To avoid this push in case the schema change is successful, we plug into
the second part of the proposal, the one which delays switching
coordinator-side schema. Once the coordinator finishes the barrier, it
marks replicas which responded as having received that schema version, and
push will be avoided for them.

If we don't know, we send it. If we see schema agreement, we can assume
> that the other side has the current version, so that after restart we don't
> need to resend needlessly.
>
> The fact that schema agreement happens lags behind actual schema
> propagation. This will case redundant schema sending. Depending on how
> heavy it is we may say it is not a problem. Schema agreement also does not
> guaranty that two nodes use the same schema. Other node may already moved
> on to a different schema.
>
In this scenario the new optimization will not be effective, we will behave
the same way as today (so not worse), the node which is ahead will make the
soft schema pull.

> That doesn't involve merging with schema tables, so is relatively fast.
> But before doing that, I think we should do the other proposal mentioned in
> this issue, which hides the latency of the heavy schema pull which merges
> schema definition with schema tables: #2109
> <https://github.com/scylladb/scylladb/issues/2109> (comment) <#2109
> (comment)
> <https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635>>
> This can also be implemented outside raft, and should work with both old
> and new mechanism. So you propose to do a global barrier (contact all nods)
> after the raft. Fine, but them it has the same issue as waiting in the
> drover. Unless you require all nodes to be alive during schema changed
> there will be races.
> What's the problem?
>
> What do you mean? What is the race? A node A may think that a node B is
> dead and not wait for it in the barrier, but B becomes alive just after the
> barrier completes.
>
Yes, this scenario will not benefit from the optimization. We could think
how to fix that case too, but it still makes sense to start with optimizing
the happy path.

**Comment by tgrabiec** [comment]
On Wed, Nov 23, 2022 at 10:29 AM Gleb Natapov ***@***.***>
wrote:

> On Tue, Nov 22, 2022 at 06:56:35AM -0800, Tomasz Grabiec wrote:
> > On Tue, Nov 22, 2022 at 2:36 PM Gleb Natapov ***@***.***>
> > wrote:
> >
> > > On Mon, Nov 21, 2022 at 1:17 PM Gleb Natapov *@*.***> wrote:
> > > @gleb-cloudius <https://github.com/gleb-cloudius>
> > > https://github.com/gleb-cloudius There are two ideas presented here.
> > > Remember that there are two kinds of schema pull, one which changes the
> > > node's local schema, and one which only pulls the definitions which
> live as
> > > a schema version in memory. The first idea with the queues does not
> plug
> > > into how schema changes are made, it only affects the request path so
> that
> > > it doesn't have to pull the schema version (in-memory one). Each
> request
> > > can use a different version of the schema. E.g. before sending a write
> > > request to some replica, we would send relevant schema version to that
> > > replica first, if it wasn't already. How do you know it was not? It is
> sent
> > > over raft. You do not know who got it and who did not (without asking).
> > >
> > > We can keep track of it in memory.
> > >
> > > Not without access to the raft internals which I am firmly against of.
> And
> > > even then the raft not always knows (but that involves leader change).
> > >
> > I think we don't need to touch raft internals, all can happen above it.
> > Firstly, the mapping doesn't have to be accurate, worst case the replica
> > will request the schema version like today. So we don't even need to
> update
> > the mapping when a schema change is made. The read/write coordinator will
> > push the schema version to target replicas with the first request which
> > uses the new version. That part doesn't involve raft.
> >
>
> So this is basically what is done now (without raft) by a schema change
> coordinator. DEFINITIONS_UPDATE and read/write verbs goes through
> the same connection, so all read/write initiate by this coordinator
> suppose to arrive after update is processed (reordering is possible
> but very unlikely).


No, DEFINITIONS_UPDATE is the hard schema push (involves a merge with
schema_tables).

I'm talking about the soft schema pull, GET_SCHEMA_VERSION. It will also
become a push after this proposal.

DEFINITIONS_UPDATE takes a long time to process, but GET_SCHEMA_VERSION
involves only in-memory deserialization, so can be processed synchronously
on the connection.



> You propose the same to be done by all nodes, not
> only schema change coordinator? Basically what it means is that each
> node, after getting new schema, will broadcast it to all other nodes.
> Seams excessive.
>

I wasn't proposing that, but note that this is what already happens today.
When a replica receives a read/write request which uses unknown schema,
that replica will pull the full schema definition. It is excessive indeed,
and that's why we need the second part which delays switching coordinators
to the new schema.


>
> > To avoid this push in case the schema change is successful, we plug into
> > the second part of the proposal, the one which delays switching
> > coordinator-side schema. Once the coordinator finishes the barrier, it
> > marks replicas which responded as having received that schema version,
> and
> > push will be avoided for them.
> >
>
> What do you call "the coordinator" here?


schema change coordinator.


> If this is schema change
> coordinator then we already send schema before new requests on the
> same channel and I do not see why we even have the problem.


DEFINITIONS_UPDATE can take a long time to process in do_merge_schema(),
and even though it is sent on the same connection, it is processed in the
background, concurrently with later requests. If later requests start using
the new schema until this completes, they will trigger both soft and hard
pull. This will cause excessive pulls, and more importantly, add latency to
those requests (twice the duration it takes to process the merge, because
we serialize schema merge).


>
> > If we don't know, we send it. If we see schema agreement, we can assume
> > > that the other side has the current version, so that after restart we
> don't
> > > need to resend needlessly.
> > >
> > > The fact that schema agreement happens lags behind actual schema
> > > propagation. This will case redundant schema sending. Depending on how
> > > heavy it is we may say it is not a problem. Schema agreement also does
> not
> > > guaranty that two nodes use the same schema. Other node may already
> moved
> > > on to a different schema.
> > >
> > In this scenario the new optimization will not be effective, we will
> behave
> > the same way as today (so not worse), the node which is ahead will make
> the
> > soft schema pull.
> >
> > > That doesn't involve merging with schema tables, so is relatively fast.
> > > But before doing that, I think we should do the other proposal
> mentioned in
> > > this issue, which hides the latency of the heavy schema pull which
> merges
> > > schema definition with schema tables: #2109
> > > <https://github.com/scylladb/scylladb/issues/2109> (comment) <#2109
> > > (comment)
> > > <
> https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635>>
> > > This can also be implemented outside raft, and should work with both
> old
> > > and new mechanism. So you propose to do a global barrier (contact all
> nods)
> > > after the raft. Fine, but them it has the same issue as waiting in the
> > > drover. Unless you require all nodes to be alive during schema changed
> > > there will be races.
> > > What's the problem?
> > >
> > > What do you mean? What is the race? A node A may think that a node B is
> > > dead and not wait for it in the barrier, but B becomes alive just
> after the
> > > barrier completes.
> > >
> > Yes, this scenario will not benefit from the optimization. We could think
> > how to fix that case too, but it still makes sense to start with
> optimizing
> > the happy path.
> >
>
> So in short, there is no solution, only a way to reduce the problem and
> it is not clear to me how serious the problem is in reality.
>

There's no perfect solution, but a solution which works in the common case.

**Comment by gleb-cloudius** [comment]
On Wed, Nov 23, 2022 at 12:23 PM Tomasz Grabiec ***@***.***>
wrote:

> On Wed, Nov 23, 2022 at 10:29 AM Gleb Natapov ***@***.***>
> wrote:
>
> > On Tue, Nov 22, 2022 at 06:56:35AM -0800, Tomasz Grabiec wrote:
> > > On Tue, Nov 22, 2022 at 2:36 PM Gleb Natapov ***@***.***>
> > > wrote:
> > >
> > > > On Mon, Nov 21, 2022 at 1:17 PM Gleb Natapov *@*.***> wrote:
> > > > @gleb-cloudius <https://github.com/gleb-cloudius>
> > > > https://github.com/gleb-cloudius There are two ideas presented here.
> > > > Remember that there are two kinds of schema pull, one which changes
> the
> > > > node's local schema, and one which only pulls the definitions which
> > live as
> > > > a schema version in memory. The first idea with the queues does not
> > plug
> > > > into how schema changes are made, it only affects the request path so
> > that
> > > > it doesn't have to pull the schema version (in-memory one). Each
> > request
> > > > can use a different version of the schema. E.g. before sending a
> write
> > > > request to some replica, we would send relevant schema version to
> that
> > > > replica first, if it wasn't already. How do you know it was not? It
> is
> > sent
> > > > over raft. You do not know who got it and who did not (without
> asking).
> > > >
> > > > We can keep track of it in memory.
> > > >
> > > > Not without access to the raft internals which I am firmly against
> of.
> > And
> > > > even then the raft not always knows (but that involves leader
> change).
> > > >
> > > I think we don't need to touch raft internals, all can happen above it.
> > > Firstly, the mapping doesn't have to be accurate, worst case the
> replica
> > > will request the schema version like today. So we don't even need to
> > update
> > > the mapping when a schema change is made. The read/write coordinator
> will
> > > push the schema version to target replicas with the first request which
> > > uses the new version. That part doesn't involve raft.
> > >
> >
> > So this is basically what is done now (without raft) by a schema change
> > coordinator. DEFINITIONS_UPDATE and read/write verbs goes through
> > the same connection, so all read/write initiate by this coordinator
> > suppose to arrive after update is processed (reordering is possible
> > but very unlikely).
>
>
> No, DEFINITIONS_UPDATE is the hard schema push (involves a merge with
> schema_tables).
>
> I'm talking about the soft schema pull, GET_SCHEMA_VERSION. It will also
> become a push after this proposal.
>
> DEFINITIONS_UPDATE takes a long time to process, but GET_SCHEMA_VERSION
> involves only in-memory deserialization, so can be processed synchronously
> on the connection.
>

Ok, so we can include soft data into the  DEFINITIONS_UPDATE, but I still
do not see how it will help.
See below.


>
>
> > You propose the same to be done by all nodes, not
> > only schema change coordinator? Basically what it means is that each
> > node, after getting new schema, will broadcast it to all other nodes.
> > Seams excessive.
> >
>
> I wasn't proposing that, but note that this is what already happens today.
> When a replica receives a read/write request which uses unknown schema,
> that replica will pull the full schema definition. It is excessive indeed,
> and that's why we need the second part which delays switching coordinators
> to the new schema.
>
>
> >
> > > To avoid this push in case the schema change is successful, we plug
> into
> > > the second part of the proposal, the one which delays switching
> > > coordinator-side schema. Once the coordinator finishes the barrier, it
> > > marks replicas which responded as having received that schema version,
> > and
> > > push will be avoided for them.
> > >
> >
> > What do you call "the coordinator" here?
>
>
> schema change coordinator.
>
>
>
Then I do not see how it is enough. Another node may start issuing requests
using new schema
and other nodes will have to pull.


> > If this is schema change
> > coordinator then we already send schema before new requests on the
> > same channel and I do not see why we even have the problem.
>
>
> DEFINITIONS_UPDATE can take a long time to process in do_merge_schema(),
> and even though it is sent on the same connection, it is processed in the
> background, concurrently with later requests. If later requests start using
> the new schema until this completes, they will trigger both soft and hard
> pull. This will cause excessive pulls, and more importantly, add latency to
> those requests (twice the duration it takes to process the merge, because
> we serialize schema merge).
>

OK. Got that. Including a soft version into the update should solve this.
But there are still other request coordinators to consider.

>
> > > Yes, this scenario will not benefit from the optimization. We could
> think
> > > how to fix that case too, but it still makes sense to start with
> > optimizing
> > > the happy path.
> > >
> >
> > So in short, there is no solution, only a way to reduce the problem and
> > it is not clear to me how serious the problem is in reality.
> >
>
> There's no perfect solution, but a solution which works in the common case.
>

I am not necessarily  sure there is no perfect solution. Maybe we just have
not seen one yet. But more importantly
to claim that the optimization is worth implementing we need to see where
the latency is really coming from. It may well
be that it comes from cases not covered by the proposal. Again, I still do
not understand what happens to requests started
not by schema change coordinator.

**Comment by kostja** [comment]
> So in short, there is no solution, only a way to reduce the problem and
> it is not clear to me how serious the problem is in reality.

 I think this comment by the submitter nails it. We should not be working on this. Raft is actively pushing schema and improves the condition quite a bit (although doesn't resolve it completely). I'd close it as "won't happen". If left open, this issue should sit in the "wish list" queue and wait for its priority to change - or it being addressed along with a more major bug or feature.

**Comment by kostja** [comment]
As a general guideline I think we should avoid assigning issues to a milestone if they don't have a test case and are not affecting any customer. Lack of a test which would allow us to accurately estimate the impact of the problem is a good indication the issue is not ready to be triaged: we don't know the cost of fixing it or the impact on the community, and without knowing these we can't prioritize the item among other things.

**Comment by tgrabiec** [comment]
This problem is affecting a customer, see the issues referencing this.

**Comment by tgrabiec** [comment]
On Wed, Nov 23, 2022 at 12:45 PM Gleb Natapov ***@***.***>
wrote:

> On Wed, Nov 23, 2022 at 12:23 PM Tomasz Grabiec ***@***.***>
> wrote:
>
> > On Wed, Nov 23, 2022 at 10:29 AM Gleb Natapov ***@***.***>
> > wrote:
> >
> > > On Tue, Nov 22, 2022 at 06:56:35AM -0800, Tomasz Grabiec wrote:
> > > > On Tue, Nov 22, 2022 at 2:36 PM Gleb Natapov ***@***.***>
> > > > wrote:
> > > >
> > > > > On Mon, Nov 21, 2022 at 1:17 PM Gleb Natapov *@*.***> wrote:
> > > > > @gleb-cloudius <https://github.com/gleb-cloudius>
> > > > > https://github.com/gleb-cloudius There are two ideas presented
> here.
> > > > > Remember that there are two kinds of schema pull, one which changes
> > the
> > > > > node's local schema, and one which only pulls the definitions which
> > > live as
> > > > > a schema version in memory. The first idea with the queues does not
> > > plug
> > > > > into how schema changes are made, it only affects the request path
> so
> > > that
> > > > > it doesn't have to pull the schema version (in-memory one). Each
> > > request
> > > > > can use a different version of the schema. E.g. before sending a
> > write
> > > > > request to some replica, we would send relevant schema version to
> > that
> > > > > replica first, if it wasn't already. How do you know it was not? It
> > is
> > > sent
> > > > > over raft. You do not know who got it and who did not (without
> > asking).
> > > > >
> > > > > We can keep track of it in memory.
> > > > >
> > > > > Not without access to the raft internals which I am firmly against
> > of.
> > > And
> > > > > even then the raft not always knows (but that involves leader
> > change).
> > > > >
> > > > I think we don't need to touch raft internals, all can happen above
> it.
> > > > Firstly, the mapping doesn't have to be accurate, worst case the
> > replica
> > > > will request the schema version like today. So we don't even need to
> > > update
> > > > the mapping when a schema change is made. The read/write coordinator
> > will
> > > > push the schema version to target replicas with the first request
> which
> > > > uses the new version. That part doesn't involve raft.
> > > >
> > >
> > > So this is basically what is done now (without raft) by a schema change
> > > coordinator. DEFINITIONS_UPDATE and read/write verbs goes through
> > > the same connection, so all read/write initiate by this coordinator
> > > suppose to arrive after update is processed (reordering is possible
> > > but very unlikely).
> >
> >
> > No, DEFINITIONS_UPDATE is the hard schema push (involves a merge with
> > schema_tables).
> >
> > I'm talking about the soft schema pull, GET_SCHEMA_VERSION. It will also
> > become a push after this proposal.
> >
> > DEFINITIONS_UPDATE takes a long time to process, but GET_SCHEMA_VERSION
> > involves only in-memory deserialization, so can be processed
> synchronously
> > on the connection.
> >
>
> Ok, so we can include soft data into the DEFINITIONS_UPDATE, but I still
> do not see how it will help.
> See below.
>
>
> >
> >
> > > You propose the same to be done by all nodes, not
> > > only schema change coordinator? Basically what it means is that each
> > > node, after getting new schema, will broadcast it to all other nodes.
> > > Seams excessive.
> > >
> >
> > I wasn't proposing that, but note that this is what already happens
> today.
> > When a replica receives a read/write request which uses unknown schema,
> > that replica will pull the full schema definition. It is excessive
> indeed,
> > and that's why we need the second part which delays switching
> coordinators
> > to the new schema.
> >
> >
> > >
> > > > To avoid this push in case the schema change is successful, we plug
> > into
> > > > the second part of the proposal, the one which delays switching
> > > > coordinator-side schema. Once the coordinator finishes the barrier,
> it
> > > > marks replicas which responded as having received that schema
> version,
> > > and
> > > > push will be avoided for them.
> > > >
> > >
> > > What do you call "the coordinator" here?
> >
> >
> > schema change coordinator.
> >
> >
> >
> Then I do not see how it is enough. Another node may start issuing requests
> using new schema
> and other nodes will have to pull.
>

The read/write coordinator (all of them) will start using the new schema
only after the barrier succeeds or times out.
This is because the schema change coordinator switches coordinators on
other nodes after the barrier.

If the barrier succeeds, it means all replicas are aware of the new schema,
so they will not pull.
If the barrier times out, we're back to current behavior.

See https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635



> I am not necessarily sure there is no perfect solution. Maybe we just have
> not seen one yet.


Yes, I meant that there is no proposal of a perfect solution.



> But more importantly
> to claim that the optimization is worth implementing we need to see where
> the latency is really coming from. It may well
> be that it comes from cases not covered by the proposal.


The latency comes from do_merge_schema() taking a long time, and the fact
that each request serializes with it via get_schema_for_write()

**Comment by gleb-cloudius** [comment]
On Wed, Nov 23, 2022 at 09:17:14AM -0800, Tomasz Grabiec wrote:
> > > > > To avoid this push in case the schema change is successful, we plug
> > > into
> > > > > the second part of the proposal, the one which delays switching
> > > > > coordinator-side schema. Once the coordinator finishes the barrier,
> > it
> > > > > marks replicas which responded as having received that schema
> > version,
> > > > and
> > > > > push will be avoided for them.
> > > > >
> > > >
> > > > What do you call "the coordinator" here?
> > >
> > >
> > > schema change coordinator.
> > >
> > >
> > >
> > Then I do not see how it is enough. Another node may start issuing requests
> > using new schema
> > and other nodes will have to pull.
> >
> 
> The read/write coordinator (all of them) will start using the new schema
> only after the barrier succeeds or times out.
> This is because the schema change coordinator switches coordinators on
> other nodes after the barrier.
> 
> If the barrier succeeds, it means all replicas are aware of the new schema,
> so they will not pull.
> If the barrier times out, we're back to current behavior.
> 
> See https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635
> 
So this algorithm has two global barriers. One is to make sure all nodes gets
the schema and another is to tell them to switch to the new one. What
happens if the second one fails on some nodes?

> 
> 
> > I am not necessarily sure there is no perfect solution. Maybe we just have
> > not seen one yet.
> 
> 
> Yes, I meant that there is no proposal of a perfect solution.
> 
> 
> 
> > But more importantly
> > to claim that the optimization is worth implementing we need to see where
> > the latency is really coming from. It may well
> > be that it comes from cases not covered by the proposal.
> 
> 
> The latency comes from do_merge_schema() taking a long time, and the fact
> that each request serializes with it via get_schema_for_write()
> 

It looks like for that it would be enough to add soft schema to the
UPDATE verb, no?

--
			Gleb.

**Comment by tgrabiec** [comment]
On Thu, Nov 24, 2022 at 9:18 AM Gleb Natapov ***@***.***>
wrote:

> On Wed, Nov 23, 2022 at 09:17:14AM -0800, Tomasz Grabiec wrote:
> > > > > > To avoid this push in case the schema change is successful, we
> plug
> > > > into
> > > > > > the second part of the proposal, the one which delays switching
> > > > > > coordinator-side schema. Once the coordinator finishes the
> barrier,
> > > it
> > > > > > marks replicas which responded as having received that schema
> > > version,
> > > > > and
> > > > > > push will be avoided for them.
> > > > > >
> > > > >
> > > > > What do you call "the coordinator" here?
> > > >
> > > >
> > > > schema change coordinator.
> > > >
> > > >
> > > >
> > > Then I do not see how it is enough. Another node may start issuing
> requests
> > > using new schema
> > > and other nodes will have to pull.
> > >
> >
> > The read/write coordinator (all of them) will start using the new schema
> > only after the barrier succeeds or times out.
> > This is because the schema change coordinator switches coordinators on
> > other nodes after the barrier.
> >
> > If the barrier succeeds, it means all replicas are aware of the new
> schema,
> > so they will not pull.
> > If the barrier times out, we're back to current behavior.
> >
> > See
> https://github.com/scylladb/scylladb/issues/2109#issuecomment-840041635
> >
> So this algorithm has two global barriers. One is to make sure all nodes
> gets
> the schema and another is to tell them to switch to the new one. What
> happens if the second one fails on some nodes?
>

If we can't contact some nodes, after a timeout we'll give up, and switch
all coordinators to the new schema (replica-side schema) anyway. All
coordinators, even those which we can't contact, will switch at the same
time because we will arm them with a specific time point before starting
the change. This arming will also ensure that we don't end up with a
disagreement if some nodes are unreachable from the beginning.


>
> >
> >
> > > I am not necessarily sure there is no perfect solution. Maybe we just
> have
> > > not seen one yet.
> >
> >
> > Yes, I meant that there is no proposal of a perfect solution.
> >
> >
> >
> > > But more importantly
> > > to claim that the optimization is worth implementing we need to see
> where
> > > the latency is really coming from. It may well
> > > be that it comes from cases not covered by the proposal.
> >
> >
> > The latency comes from do_merge_schema() taking a long time, and the fact
> > that each request serializes with it via get_schema_for_write()
> >
>
> It looks like for that it would be enough to add soft schema to the
> UPDATE verb, no?
>

get_schema_for_write() also needs to wait for the hard merge, so that
writes which depend on the new schema (e.g. new columns) can actually be
stored.

**Comment by tgrabiec** [comment]
With RAFT, we could avoid the problem to a large extent by making commands apply only to the in-memory schema, and persist them to schema tables when snapshotting is done, in the background.

**Comment by mykaul** [comment]
> With RAFT, we could avoid the problem to a large extent by making commands apply only to the in-memory schema, and persist them to schema tables when snapshotting is done, in the background.

Adding raft keyword just so we'll remember to look at this when working on raft tests.

**Comment by kbr-scylla** [comment]
I think there's no point implementing this as long as the old schema pull code lives and schema propagation without Raft is still possible. Which it is because of gossip-based RECOVERY mode. Therefore, I will consider this issue BLOCKED until https://github.com/scylladb/scylladb/issues/18791 and https://github.com/scylladb/scylladb/issues/15422 are fixed.
